### PR TITLE
Ibc cosmwasm contracts

### PR DESCRIPTION
**Pull request by nik-suri**
#2198 #2199 

This PR adds the two new cosmwasm contract for IBC generic message emission: `wormchain-ibc-receiver` and `wormhole-ibc`. These two contracts implement a custom IBC protocol. The `wormhole-ibc` contract should be deployed to IBC-enabled chains, and the various `wormhole-ibc` contract should connect via IBC to the `wormchain-ibc-receiver` contract, which will live on Wormchain.

For existing Cosmos deployments, the `wormhole` cosmwasm contract should be migrated to the `wormhole-ibc` cosmwasm contract after it is uploaded to the chain.

### CODE REVIEW

File: cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    if let Some(counter_version) = msg.counterparty_version() {
+        if counter_version != IBC_APP_VERSION {
+            return Err(StdError::generic_err(format!(
+                "Counterparty version must be `{}`",
+                IBC_APP_VERSION
+            )));
+        }
+    }
+
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs:41]: Should we be checking CHAIN_CONNECTIONS somewhere in here?  Otherwise it doesn't look like that map is used anywhere.

File: cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        let packet = msg.packet;
+        // which local channel did this packet come on
+        let channel_id = packet.dest.channel_id;
+        let wormhole_msg: WormholeIbcPacketMsg = from_slice(&packet.data)?;
+        match wormhole_msg {
+            WormholeIbcPacketMsg::Publish { msg: publish_msg } => {
+                receive_publish(deps, channel_id, publish_msg)
+            }
+        }
+    })()
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs:97]: style nit: I'd prefer if we move this out into a named function (like `handle_ibc_packet`) rather than inlining it as a closure here.

File: cosmwasm/contracts/wormchain-ibc-receiver/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+pub enum WormholeIbcPacketMsg {
+    Publish { msg: Response },
+}
+
+/// Contract queries
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum QueryMsg {
+    #[returns(AllChainConnectionsResponse)]
+    AllChainConnections {},
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/msg.rs:26]: nit:

```suggestion
    AllChainConnections,
```

File: cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
+use anyhow::{ensure, Context};
+use cosmwasm_std::{entry_point, to_binary, Binary, Deps, Empty, Event, StdError, StdResult};
+use cosmwasm_std::{DepsMut, Env, MessageInfo, Order, Response};
+use cw2::{get_contract_version, set_contract_version};
+use semver::Version;
+use serde_wormhole::RawMessage;
+use wormhole::ibc_receiver::{Action, GovernancePacket};
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs:7]: I think we're missing the rust SDK changes that add these types.

Comment (nik-suri) [reply cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs:7]: This PR is meant to be merged into the `ibc-sdk-updates` branch. Here's the PR for this branch, which should be merged into `main` first: https://github.com/wormhole-foundation/wormhole/pull/2590

File: cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs (hunk 1 above)
Comment (nik-suri) [reply cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs:41]: See the whitepaper PR for how the CHAIN_CONNECTIONS map should be used.

TLDR - it's a mapping that we need to maintain somewhere for security, so that the guardian watcher can verify where IBC messages come from. @bruce-riley @evan-gray @hendrikhofstadt and I decided to keep it in this ibc-wormchain-receiver contract, and only allow updates through governance VAAs.

File: cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs (hunk 2 above)
Comment (nik-suri) [reply cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs:97]: can do. Any particular reason why you feel that would be better style?

File: cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        deps.as_mut(),
+        mock_env(),
+        execute_info,
+        ExecuteMsg::PostMessage {
+            message: wh_message,
+            nonce: 1,
+        },
+    )
+    .unwrap();
+    assert_eq!(res.attributes.len(), 5);
```
Comment (hendrikhofstadt) [inline cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs:59]: any way to check for the IBC packet?

**Review by hendrikhofstadt — CHANGES_REQUESTED** [review]
See inline comments

File: cosmwasm/contracts/wormhole-ibc/src/ibc.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+/// Never should be called as the other side never sends packets
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_receive(
+    _deps: DepsMut,
+    _env: Env,
+    _msg: IbcPacketReceiveMsg,
+) -> StdResult<IbcReceiveResponse> {
+    Ok(IbcReceiveResponse::new()
+        .set_ack(b"{}")
+        .add_attribute("action", "ibc_packet_ack"))
```
[the remark this answers is not in the document]
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/ibc.rs:91]: we don't need to send any data in the acknowledgement, which is why I use the empty byte string `b"{}"`

File: cosmwasm/contracts/wormhole-ibc/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        Version::parse(CONTRACT_VERSION).context("could not parse new contract version")?;
+    if saved_version >= new_version {
+        return Err(StdError::generic_err("Cannot upgrade from a newer version").into());
+    }
+
+    // set the new version
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // call the core contract migrate function
+    wormhole::contract::migrate(deps, env, msg).context("wormhole core migration failed")
```
[the remark this answers is not in the document]
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/contract.rs:61]: The core contract migrate function is actually empty. Do we have to call the core contract migrate function here?

File: cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    _deps: DepsMut,
+    _env: Env,
+    _msg: IbcPacketTimeoutMsg,
+) -> StdResult<IbcBasicResponse> {
+    Ok(IbcBasicResponse::new().add_attribute("action", "ibc_packet_timeout"))
+}
+
+#[cfg(test)]
+mod tests {
+    // test opening channel
```
[the remark this answers is not in the document]
Comment (nik-suri) [reply cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs:152]: yes

File: cosmwasm/contracts/wormhole-ibc/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+fn post_message_ibc(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> anyhow::Result<Response> {
+    // search for a channel bound to counterparty with the port "wasm.<wormchain_addr>"
+    let ibc_channels = deps
+        .querier
+        .query::<ListChannelsResponse>(&IbcQuery::ListChannels { port_id: None }.into())
```
[the remark this answers is not in the document]
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/contract.rs:87]: good point, I'll update to use the rust iterator functions instead.

File: cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        .context("failed to set contract version")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("owner", info.sender)
+        .add_attribute("version", CONTRACT_VERSION))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: Empty) -> Result<Response, anyhow::Error> {
```
Comment (evan-gray) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs:37]: Where is this migrate logic from? It does not match how we perform migrates in the wormhole contract, and more specifically blocks upgrading from the wormhole contract. All of this versioning code appears new / different and causes this issue.

**Review by evan-gray — CHANGES_REQUESTED** [review]
When @panoel attempted to upgrade the `wormhole` contract, it failed due to the new versioning checks.

File: cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs (hunk 10 above)
Comment (nik-suri) [reply cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs:37]: I got it from here -- https://medium.com/cosmwasm/cosmwasm-for-ctos-ii-advanced-usage-ee04ce95d1d0

When we run the migrate from `wormhole` to `wormhole-ibc`, I was under the impression that only the migrate function in the `wormhole` contract would be called and this would pass.

I can remove these checks from both contracts though.

File: cosmwasm/contracts/wormhole-ibc/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    let channel_id = ibc_channels
+        .iter()
+        .find(|c| c.counterparty_endpoint.port_id == WORMCHAIN_IBC_RECEIVER_PORT)
+        .map(|c| c.endpoint.channel_id.clone())
+        .context(
+            "no channel connecting to wormchain contract port {WORMCHAIN_IBC_RECEIVER_PORT}",
+        )?;
+
+    // compute the packet timeout (infinite timeout)
+    let packet_timeout = env.block.time.plus_seconds(PACKET_LIFETIME).into();
```
[the remark this answers is not in the document]
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/contract.rs:100]: IBC message wouldn’t send with a timeout of 0 — execution failed with the following message
```
raw_log: 'failed to execute message; message index: 0: dispatch: submessages: packet
  failed basic validation: packet timeout height and packet timeout timestamp cannot
  both be 0: invalid packet'
```

File: cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    // match the governance action and execute the corresponding logic
+    match govpacket.action {
+        Action::UpdateChannelChain {
+            channel_id,
+            chain_id,
+        } => {
+            ensure!(chain_id != Chain::Wormchain, "the wormchain-ibc-receiver contract should not maintain channel mappings to wormchain");
+
+            let channel_id_str = String::from_utf8(channel_id.to_vec())
+                .context("failed to parse channel-id as utf-8")?;
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormchain-ibc-receiver/src/contract.rs:113]: nit: use `str::from_utf8`

File: cosmwasm/contracts/wormhole-ibc/src/contract.rs [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+    res = match tx_index {
+        Some(index) => res.add_attribute("message.tx_index", index.to_string()),
+        None => res,
+    };
+    res = res.add_attribute("message.block_height", block_height);
+
+    // Send the result attributes over IBC on this channel
+    let packet = WormholeIbcPacketMsg::Publish {
+        msg: res.attributes.clone(),
```
Comment (jynnantonix) [inline cosmwasm/contracts/wormhole-ibc/src/contract.rs:170]: This clone looks unnecessary?

File: cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs
Comment (jynnantonix) [inline cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs:1]: Can we either enable these tests or delete this file?

**Review by jynnantonix — DISMISSED** [review]

File: cosmwasm/contracts/wormhole-ibc/src/contract.rs (hunk 13 above)
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/contract.rs:170]: Without it, `res` is partially moved. Since we return `res` at the end of the function and still want to emit the same attributes as part of the event, we need the clone here.

File: cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs
Comment (nik-suri) [reply cosmwasm/contracts/wormhole-ibc/src/contract/tests.rs:1]: Deleted.

**Review by jynnantonix — APPROVED** [review]
